core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/async/SplittingTransformer.java

```diff
@@ -314,14 +323,23 @@ public void onStream(SdkPublisher<ByteBuffer> publisher) {
                     );
                 }
             }
-            publisher.subscribe(new IndividualPartSubscriber<>(this.individualFuture, response));
+
+            if (!resultFuture.isDone()) {
+                CompletableFutureUtils.forwardResultTo(upstreamFuture, resultFuture);
+            }
+
+            publisher.subscribe(new IndividualPartSubscriber<>(this.individualFuture, response,
+                                                               isFirstIndividualSubscriber.compareAndSet(true, false)));
         }
 
         @Override
         public void exceptionOccurred(Throwable error) {
-            publisherToUpstream.error(error);
             log.trace(() -> "calling exceptionOccurred on the upstream transformer");
             upstreamResponseTransformer.exceptionOccurred(error);
+
+            if (!isFirstPart || onStreamCalled.get()) {
```

Comment:
Can we add some comments explaining this behavior since it's not obvious from the code?